zrml/neo-swaps/src/types/liquidity_tree.rs

```diff
@@ -0,0 +1,1205 @@
+///
+/// # Generics
+///
+/// - `D`: A getter for the depth of the tree.
+pub struct LiquidityTreeMaxNodes<D>(PhantomData<D>);
+
+impl<D> Get<u32> for LiquidityTreeMaxNodes<D>
+where
+    D: Get<u32>,
+{
+    fn get() -> u32 {
+        2u32.saturating_pow(D::get() + 1).saturating_sub(1)
+    }
+}
+
+/// A segment tree used to track balances of liquidity shares which allows `O(log(n))` distribution
+/// of fees.
+///
+/// Each liquidity provider owns exactly one node of the tree which records their stake and fees.
+/// When a liquidity provider leaves the tree, the node is not removed from the tree, but marked as
+/// _abandoned_ instead. Abandoned nodes are reassigned when new LPs enter the tree. Nodes are added
+/// to the leaves of the tree only if there are no abandoned nodes to reassign.
+///
+/// Fees are lazily propagated down the tree. This allows fees to be deposited to the tree in `O(1)`
+/// (fees deposited at the root and later propagated down). If a particular node requires to know
+/// what its fees are, propagating fees to this node takes `O(depth)` operations (or, equivalently,
+/// `O(log_2(node_count))`).
+///
+/// # Attributes
+///
+/// - `nodes`: A vector which holds the nodes of the tree. The nodes are ordered by depth (the root
+///   is the first element of `nodes`) and from left to right. For example, the right-most
+///   grandchild of the root is at index `6`.
+/// - `account_to_index`: Maps an account to the node that belongs to it.
+/// - `abandoned_nodes`: A vector that contains the indices of abandoned nodes.
+///
+/// # Generics
+///
+/// - `T`: The pallet configuration.
+/// - `U`: A getter for the maximum depth of the tree. Using a depth larger than `31` will result in
+///   undefined behavior.
+#[derive(Decode, Encode, Eq, MaxEncodedLen, RuntimeDebugNoBound, TypeInfo)]
+#[scale_info(skip_type_params(T, U))]
+pub struct LiquidityTree<T, U>
+where
+    T: Config,
+    U: Get<u32>,
+{
+    pub(crate) nodes: BoundedVec<Node<T>, LiquidityTreeMaxNodes<U>>,
+    pub(crate) account_to_index: BoundedBTreeMap<T::AccountId, u32, LiquidityTreeMaxNodes<U>>,
+    pub(crate) abandoned_nodes: BoundedVec<u32, LiquidityTreeMaxNodes<U>>,
+}
+
+// Boilerplate implementation because Rust is confused by the generic parameter `U`.
+impl<T, U> Clone for LiquidityTree<T, U>
+where
+    T: Config,
+    U: Get<u32>,
+{
+    fn clone(&self) -> Self {
+        LiquidityTree {
+            nodes: self.nodes.clone(),
+            account_to_index: self.account_to_index.clone(),
+            abandoned_nodes: self.abandoned_nodes.clone(),
+        }
+    }
+}
+
+// Boilerplate implementation because Rust is confused by the generic parameter `U`.
+impl<T, U> PartialEq for LiquidityTree<T, U>
+where
+    T: Config,
+    U: Get<u32>,
+{
+    fn eq(&self, other: &Self) -> bool {
+        self.nodes == other.nodes
+            && self.account_to_index == other.account_to_index
+            && self.abandoned_nodes == other.abandoned_nodes
+    }
+}
```

Comment:
> Instead of using U: Get<u32> you could also just use the generic without a trait bound.

`LiquidityTreeMaxNodes` requires that trait bound.

> Maybe that helps, I don't know.

So you didn't check if this compiles?

Reply:
> So you didn't check if this compiles?

Correct. I am usually not in the code editor when reviewing.

But I checked it now, it should work.

5f517ecf91c6aa65c291524faed5b7b13d9ed29e

Reply:
> https://github.com/zeitgeistpm/zeitgeist/commit/5f517ecf91c6aa65c291524faed5b7b13d9ed29e

Cool. 👍

> Correct. I am usually not in the code editor when reviewing.

Not sure I follow. But pushing into features branches that you're not the author of can cause various workflow distruptions, so I kindly ask not to do that.